File: src/main/java/org/foolish/ubc/FiroeWithoutBraneMind.java
package org.foolish.ubc;

import org.foolish.ast.AST;

/**
 * FIR without a braneMind queue.
 * Used for simple values that don't require step-by-step evaluation.
 */
public abstract class FiroeWithoutBraneMind extends FIR {

    protected FiroeWithoutBraneMind(AST ast, String comment) {
        super(ast, comment);
    }

    protected FiroeWithoutBraneMind(AST ast) {
        this(ast, null);
    }

    /**
     * FiroeWithoutBraneMind instances are never underevaluated.
     * They represent finalized values.
     */
    @Override
    public boolean underevaluated() {
        return false;
    }
}

File: foolish-core/src/main/scala/org/foolish/fvm/scubc/FiroeWithoutBraneMind.scala
package org.foolish.fvm.scubc

import org.foolish.ast.AST

/**
 * FIR without a braneMind queue.
 * Used for simple values that don't require step-by-step evaluation.
 */
abstract class FiroeWithoutBraneMind(val ast: AST, val comment: Option[String] = None) extends FIR:

  /** FiroeWithoutBraneMind instances don't require stepping */
  def step(): Unit = () // No-op

  /** FiroeWithoutBraneMind instances are never NYE (Not Yet Evaluated) */
  def isNye: Boolean = false
